velox/exec/HashProbe.h

```diff
@@ -449,32 +453,46 @@ class HashProbe : public Operator {
   // with multiple matches.
   class LeftSemiProjectJoinTracker {
    public:
+    static const char* kPassed;
+    static const char* kNull;
+
     // Called for each row and indicates whether the filter passed or not.
     // Expects that probe side rows with multiple matches are next to each
     // other. Calls onLast just once for each probe side row.
     template <typename TOnLast>
-    void advance(vector_size_t row, bool passed, TOnLast onLast) {
+    void
+    advance(vector_size_t row, std::optional<bool> passed, TOnLast onLast) {
       if (currentRow != row) {
         if (currentRow != -1) {
-          onLast(currentRow, currentRowPassed);
+          onLast(
+              currentRow,
+              currentRowUnknown ? std::nullopt
+                                : std::optional(currentRowPassed));
         }
         currentRow = row;
         currentRowPassed = false;
```

Comment:
I used a separate boolean, because an optional required a somewhat convoluted logic. Notice that "currentRowPassed = passed" above would be incorrect. I'm changing to optional as I think every other reader will have the same question.

```
      if (passed.has_value()) {
        if (currentRowPassed.has_value()) {
          currentRowPassed = currentRowPassed.value() | passed.value();
        } else {
          currentRowPassed = passed;
        }
      }
```